router/src/main/java/io/micronaut/web/router/version/VersionedRouter.java

```diff
@@ -0,0 +1,128 @@
+package io.micronaut.web.router.version;
+
+import io.micronaut.http.HttpMethod;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.HttpStatus;
+import io.micronaut.http.filter.HttpFilter;
+import io.micronaut.web.router.RouteMatch;
+import io.micronaut.web.router.Router;
+import io.micronaut.web.router.UriRoute;
+import io.micronaut.web.router.UriRouteMatch;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Decorated router with version matching.
+ */
+public class VersionedRouter implements Router {
```

Comment:
I'm not sure it is necessary to wrap the existing router. I think it would probably be better just to fold this behaviour into `DefaultRouter`